README.md
# Task Forse

> Nel dubbio... task forse!

"Task Forse" is an independent project for monitoring of public task forces' activity during COVID-19 emergency in 2020 in Italy: https://www.taskforse.it/.

It's a no-profit and community-driven initiative supported by [Ondata](https://ondata.it) and based on a shared database hosted by [Google Sheets](https://www.google.com/intl/it/sheets/about/) and [Google Form](https://www.google.it/intl/it/forms/about/) and a web application built on top of [NextJS](https://nextjs.org/).

The name "Task Forse" is a joke in Italian language: _force_ in English and _forse_ in Italian (_maybe_ in English) sound similar, so in Italy the word _task forse_ suggests some amount of uncertainty in the management of emergency using task forces appointed by the government and local public institutions.

## Database

Shared database: [15LmCiYKg2cWzovAiqquhp_lYsaBSuGNau7suUkQddl8](https://docs.google.com/spreadsheets/d/15LmCiYKg2cWzovAiqquhp_lYsaBSuGNau7suUkQddl8/edit?usp=sharing).

Sheets:
- [Meta](https://docs.google.com/spreadsheets/d/15LmCiYKg2cWzovAiqquhp_lYsaBSuGNau7suUkQddl8/edit#gid=0): database metadata with tables and fields descriptions.
- [Task forses](https://docs.google.com/spreadsheets/d/15LmCiYKg2cWzovAiqquhp_lYsaBSuGNau7suUkQddl8/edit#gid=1428144440) (task forces): table with all monitored task forces
- [Membri](https://docs.google.com/spreadsheets/d/15LmCiYKg2cWzovAiqquhp_lYsaBSuGNau7suUkQddl8/edit#gid=886519616) (members): table with all known members which are part of at least one task force
- [Verbali](https://docs.google.com/spreadsheets/d/15LmCiYKg2cWzovAiqquhp_lYsaBSuGNau7suUkQddl8/edit#gid=939478784) (minutes): table with all official minutes produced and pyblished by task forces
- [Risorse](https://docs.google.com/spreadsheets/d/15LmCiYKg2cWzovAiqquhp_lYsaBSuGNau7suUkQddl8/edit#gid=369768687) (resources): table with all sort of external resources related to task forces' activity

You can freely access the shared database, download it in all formats supported by Google Sheet and reuse it in agreement with a [Creative Commons Attribution (CC-BY)](https://creativecommons.org/licenses/by/4.0/deed.it) license.

## Web application

Running instance: https://www.taskforse.it/.

### Source code

Public repository: https://github.com/ondata/taskforse.it.

Source code is released under a [MIT](https://opensource.org/licenses/MIT) license on Github. Contributions are welcome. Feel free to open issues and submit a pull request at any time!

### Software architecture

This application is composed by three services:
- a reverse proxy with caching as a middleware between client and nextjs application and between nextjs application and remote google sheet
- a nextjs application with SSR enabled
- an interactive API documentation

The remote database is a simple Google Sheet with public visibility accessed by nextjs application only on server-side.

### Production

You can deploy and run the application in a production environment using [Docker](https://www.docker.com/). A `docker-compose.yml` is provided to simplify deploy using [Docker Compose](https://docs.docker.com/compose/) utility.

The application is composed by 3 containers based on images hosted by [Github Packages](https://github.com/features/packages): https://github.com/ondata/taskforse.it/packages.
- [app:latest](https://github.com/ondata/taskforse.it/packages/233171): nextjs application
- [proxy:latest](https://github.com/ondata/taskforse.it/packages/233169): nginx reverse proxy with caching
- [osa:latest](https://github.com/ondata/taskforse.it/packages/233918): swagger ui for API documentation

You can simply copy and paste the [docker-compose.yml file](https://github.com/ondata/taskforse.it/blob/master/docker-compose.yml) provided and run `docker-compose up`. Then you can open `http://localhost:8080` using your favourite browser.

You can set environment variables using a `.env` file:
- PROXY_URL - URL of nginx server with path for proxied Google Sheet, default: http://proxy/db
- APP_URL - URL of nextjs application, default: http://app:3000
- OAS_URL - URL of swagger ui service, default: http://oas-ui:8080
- PROD_PORT - Port exposed on host, default: 8080

### Development

If you want to contribute to the development, you can fork and clone this repository, make your changes and finally send a pull request.

NextJS is a node application, so after cloning you must install dependencies running `npm install`.

Main dependencies:
- [Font Aawesome icons for react](https://fortawesome.com/)
- [Material-UI](https://material-ui.com/)
- [Axios](https://github.com/axios/axios)
- [Lodash](https://lodash.com/)
- [NextJS](https://nextjs.org/)
- [React](https://reactjs.org/)

You can start development server with hot reloading running `npm run dev`. You can also test the production instance running `npm run start`. Then you can open `http://localhost:3000` using your favourite browser. Note that reverse proxy and caching are unavailable outside docker containers.

### API

All data are available on a REST API (only GET verb is supported): https://www.taskforse.it/api/v1. There is also an interactive documentation powered by [Swagger UI](https://swagger.io/tools/swagger-ui/) and compliant to [OpenAPI Specification v3](http://spec.openapis.org/oas/v3.0.3): https://www.taskforse.it/oas/.

If you want to contribute to the API documentation, you can run [Swagger Editor](https://swagger.io/tools/swagger-editor/) using the docker-compose.yml file in `oas/` folder.

> Warning: [CORS](https://developer.mozilla.org/it/docs/Web/HTTP/CORS) is disabled by default for all origins, if you need to access API from a client-side application feel free to open an issue.

## License

The source code is released under the [MIT License](https://opensource.org/licenses/MIT).

The database is released under the [Creative Commons Attribution License](https://creativecommons.org/licenses/by/4.0/deed.it).


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

## v1.2.0 - 2020-05-30

Added
- REST API release
- OAS3 with Swagger UI

Changed
- Documentation

## v1.1.2 - 2020-05-21

Added
- Incremental Static Regeneration introduced in NextJS v9.4

Changed
- Dependencies update

Fixed
- Issue in number parsing

## v1.1.1 - 2020-05-19

Added
- Support to Github Packages

Changed
- About page content

Fixed
- Minor issues

## v1.1.0 - 2020-05-17

Added
- Single member page
- Ondata logo
- Paypal button

Fixed
- Minor issues

## v1.0.0 - 2020-05-12

First public release.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/) and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).


src/pages/about.js
import { NextSeo } from 'next-seo'

import {
    orderBy,
    map,
    keys,
    values,
    size,
    sumBy,
} from 'lodash'

import {
    Container,
    Typography,
    List,
    ListItem,
    ListItemText,
    Grid,
} from '@material-ui/core'

import {
    REVALIDATE_INTERVAL,
    GFORM_URL_ISSUE,
    getGFormUrl,
    getContributors,
    getContributions,
} from '../config'

import {
    Header,
    Footer,
    GridItem,
} from '../components'

export default function Index({ contributionsByContributor, totalContributionsCount }) {

    const contributors = keys(contributionsByContributor)
    const authoredContributionsCount = sumBy(values(contributionsByContributor), size)

    return (
        <>

            <NextSeo
                title="About"
                description="Task Forse è un progetto dell'associazione onData."
                openGraph={{
                    title: `About | Task Forse by @ondatait`,
                    description: "Task Forse è un progetto dell'associazione onData.",
                }}
            />

            <Header
                title="Home"
                href="/"
                as="/"
            />

            <main>

                <Container maxWidth="sm">

                    <Typography variant="h2" gutterBottom>
                        Il progetto
                    </Typography>

                    <Typography gutterBottom>
                        <strong>Task Forse</strong> è un progetto dell'<a target="_blank" href="https://ondata.it/">associazione onData</a> e ha l'<strong>obiettivo di raccogliere e integrare tutti i dati delle task force</strong> attivate per fronteggiare l'emergenza nazionale COVID-19, disperse in molte pagine web e documenti pdf difficilmente reperibili.
                    </Typography>

                    <Typography gutterBottom>
                        <strong>Uno spazio unico</strong> in cui puoi trovare l'elenco delle task force, tutti i membri che le costituiscono e tutti i documenti e le risorse che producono, con tutti i dettagli che aggiungeremo insieme nel tempo.
                    </Typography>

                    <Typography gutterBottom>
                        <a target="_blank" href="http://blog.ondata.it/nel-dubbio-task-forse/">Task Forse è un progetto di comunità</a> e una <strong>banca dati</strong> a supporto di azioni di monitoraggio delle attività espletate da questi gruppi, in modo da avere dei riferimenti diretti per eventuali campagne di richiesta di apertura dei dati.
                    </Typography>

                    <Typography gutterBottom>
                        <strong>Giovedì 21 maggio alle 12:00</strong> abbiamo presentato il progetto in un webinar in diretta su <a target="_blank" href="https://www.facebook.com/ondata.it/videos/565486467440217/">Facebook</a> e <a target="_blank" href="https://www.youtube.com/watch?v=YNKJxkpD1aQ">Youtube</a> con Alessio, Andrea, Alice, Jacopo e Lorenzo.
                    </Typography>
                    
                    <iframe width="100%" height="315" src="https://www.youtube-nocookie.com/embed/YNKJxkpD1aQ" frameBorder="0" allow="accelerometer; autoplay; encrypted-media; gyroscope; picture-in-picture" allowfullscreen></iframe>

                    <Typography variant="h2" gutterBottom>
                        onData - Associazione di promozione sociale
                    </Typography>

                    <Typography gutterBottom>
                        Siamo un'<a target="_blank" href="https://ondata.it/">associazione senza scopo di lucro</a> che promuove l'apertura dei dati pubblici per renderli accessibili a tutti. Siamo attivi sull'emergenza COVID-19 con azioni volte alla trasparenza e all'accessibilità dei dati pubblici. <a target="_blank" href="https://sostieni.ondata.it/">Sostienici anche tu!</a>
                    </Typography>

                    <form action="https://www.paypal.com/cgi-bin/webscr" method="post" target="_blank" style={{textAlign:"center"}}>
		                <input type="hidden" name="cmd" value="_s-xclick" />
		                <input type="hidden" name="hosted_button_id" value="5NVA5STRUX46Y" />
		           	    <input type="image" src="https://www.paypalobjects.com/it_IT/IT/i/btn/btn_donateCC_LG.gif" border="0" name="submit" alt="PayPal è il metodo rapido e sicuro per pagare e farsi pagare online." />
		                <img alt="" border="0" src="https://www.paypalobjects.com/it_IT/i/scr/pixel.gif" width="1" height="1" />
		            </form>

                    <Typography variant="h2" gutterBottom>
                        Crediti
                    </Typography>

                    <Typography gutterBottom>
                        Questo sito è sviluppato e mantenuto con &hearts; da <a target="_blank" href="https://github.com/jenkin">jenkin</a>, utilizzando <a target="_blank" href="https://nextjs.org/">NextJS</a> e <a target="_blank" href="https://material-ui.com/">Material-UI</a>.
                        {` `}
                        Il nome è un'idea di <a target="_blank" href="https://www.linkedin.com/in/andreaborruso/">aborruso</a>, il graphic design di <a target="_blank" href="https://www.linkedin.com/in/jacoposolmi/">japi</a> la progettazione e gestione del database di <a target="_blank" href="https://www.linkedin.com/in/lorenzo-perone-5aa8412/">lore</a> e <a target="_blank" href="https://www.linkedin.com/in/alicecorona/">alice</a>.
                    </Typography>

                    <Typography variant="h2" gutterBottom>
                        Contributori
                    </Typography>

                    <Typography gutterBottom>
                        Finora hanno contribuito all'arricchimento del database {size(contributors)} persone per un totale di {authoredContributionsCount} contributi, a cui se ne aggiungono {totalContributionsCount - authoredContributionsCount} anonimi. Grazie!
                    </Typography>

                </Container>

                <Container maxWidth="md">

                    <Grid container spacing={1}>
                        {
                            map(
                                orderBy(
                                    contributors,
                                    contributor => size(contributionsByContributor[contributor]),
                                    "desc"
                                ),
                                contributor => (
                                    <Grid item xs={12} sm={6} md={4} key={contributor}>
                                        <GridItem
                                            title={contributor}
                                            subtitle={`${size(contributionsByContributor[contributor])} contribut${size(contributionsByContributor[contributor]) > 1 ? "i" : "o"}`}
                                        />
                                    </Grid>
                                )
                            )
                        }
                    </Grid>

                </Container>

                <Container maxWidth="sm">

                    <Typography variant="h2" gutterBottom>
                        Roadmap
                    </Typography>

                    <Typography gutterBottom>
                        <List dense>
                            <ListItem>
                                <ListItemText>Progressive Web Application</ListItemText>
                            </ListItem>
                            <ListItem>
                                <ListItemText>Sitemap</ListItemText>
                            </ListItem>
                            <ListItem button component="a" target="_blank" href={getGFormUrl(GFORM_URL_ISSUE)}>
                                <ListItemText>+ Suggerisci...</ListItemText>
                            </ListItem>
                        </List>
                    </Typography>

                </Container>

            </main>

            <Footer />
        </>
    )
}

export async function getStaticProps() {
    return {
        props: {
            contributionsByContributor: await getContributors(),
            totalContributionsCount: size(await getContributions()),
        },
        unstable_revalidate: REVALIDATE_INTERVAL,
    }
}
